From bd3c363b15242cf97710fc960d8c59bab22af6b2 Mon Sep 17 00:00:00 2001
From: dianlujitao <dianlujitao@lineageos.org>
Date: Thu, 18 Jan 2018 21:24:30 +0800
Subject: [PATCH 1/4] power: Prepare for power profile support

 * Add support for system performance profiles and CPU boost.
 * Uses mpctl to configure the CPUs in the appropriate modes.
 * Implement binderized Lineage power HAL.

Author: Steve Kondik <shade@chemlab.org>
Date:   Sat May 17 03:37:53 2014 -0700

    power: Add support for CPU boost hint

    Change-Id: I07c3e8daf8a5f3b568e27334f26d4ba8f4cf40c4

Author: Steve Kondik <shade@chemlab.org>
Date:   Mon May 19 17:26:34 2014 -0700

    power: Add support for SET_PROFILE hints

     * Add support for system performance profiles.
     * Uses mpctl to configure the CPUs in the appropriate modes.
     * PERFORMANCE = all cores at max
     * POWER_SAVE = max two cores at non-turbo voltage
     * If a custom profile is set, don't honor any other hints
     * Clean up the code and firm up the locking.

    Change-Id: Ie6acada805780c9ae6e6bc2002843aef638ca63b

Author: Steve Kondik <steve@cyngn.com>
Date:   Tue Nov 3 03:27:42 2015 -0800

    power: Update for PerformanceManager changes

     * Undo damage caused by the previous patch which broke
       all hint arguments. We were actually using these wrong to
       begin with anyway.
     * Add support for get_profile
     * Clean up code

    Change-Id: Ibc3f21bfb7aa46ec97b9b63d09737d4331a5a714

Change-Id: I71357fac7b72de427d5f755f7ed572615ea11675
---
 Android.mk     |  3 ++-
 Power.cpp      | 37 +++++++++++++++++++++++++++++++++++--
 Power.h        | 11 +++++++++--
 hint-data.h    |  2 ++
 power-common.h | 11 +++++++++++
 power-helper.c | 14 ++++++++++++--
 power-helper.h |  2 ++
 service.cpp    | 11 ++++++-----
 8 files changed, 79 insertions(+), 12 deletions(-)

diff --git a/Android.mk b/Android.mk
index af45ccf..fe0f659 100644
--- a/Android.mk
+++ b/Android.mk
@@ -1,5 +1,5 @@
 # Copyright (C) 2017 The Android Open Source Project
-# Copyright (C) 2017 The LineageOS Project
+# Copyright (C) 2017-2018 The LineageOS Project
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.
@@ -137,6 +137,7 @@ LOCAL_MODULE := android.hardware.power@1.1-service-qti
 LOCAL_INIT_RC := android.hardware.power@1.1-service-qti.rc
 LOCAL_SHARED_LIBRARIES += android.hardware.power@1.1
 endif
+LOCAL_SHARED_LIBRARIES += vendor.lineage.power@1.0
 LOCAL_MODULE_TAGS := optional
 LOCAL_MODULE_OWNER := qcom
 LOCAL_VENDOR_MODULE := true
diff --git a/Power.cpp b/Power.cpp
index bb4bb50..67d495d 100644
--- a/Power.cpp
+++ b/Power.cpp
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2017 The Android Open Source Project
- * Copyright (C) 2017 The LineageOS Project
+ * Copyright (C) 2017-2018 The LineageOS Project
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -21,6 +21,8 @@
 #define LOG_TAG "android.hardware.power@1.1-service-qti"
 #endif
 
+// #define LOG_NDEBUG 0
+
 #include <android/log.h>
 #include <utils/Log.h>
 #include "Power.h"
@@ -52,6 +54,7 @@ using ::android::hardware::power::V1_1::PowerStateSubsystem;
 using ::android::hardware::hidl_vec;
 using ::android::hardware::Return;
 using ::android::hardware::Void;
+using ::android::hardware::power::V1_0::Feature;
 
 Power::Power() {
     power_init();
@@ -64,7 +67,7 @@ Return<void> Power::setInteractive(bool interactive)  {
 }
 
 Return<void> Power::powerHint(PowerHint hint, int32_t data) {
-    power_hint(static_cast<power_hint_t>(hint), data ? (&data) : NULL);
+    power_hint(static_cast<power_hint_t>(hint), &data);
     return Void();
 }
 
@@ -227,6 +230,36 @@ Return<void> Power::powerHintAsync(PowerHint hint, int32_t data) {
 }
 #endif
 
+Return<int32_t> Power::getFeature(LineageFeature feature)  {
+    if (feature == LineageFeature::SUPPORTED_PROFILES) {
+        return get_number_of_profiles();
+    }
+    return -1;
+}
+
+status_t Power::registerAsSystemService() {
+    status_t ret = 0;
+
+    ret = IPower::registerAsService();
+    if (ret != 0) {
+        ALOGE("Failed to register IPower (%d)", ret);
+        goto fail;
+    } else {
+        ALOGI("Successfully registered IPower");
+    }
+
+    ret = ILineagePower::registerAsService();
+    if (ret != 0) {
+        ALOGE("Failed to register ILineagePower (%d)", ret);
+        goto fail;
+    } else {
+        ALOGI("Successfully registered ILineagePower");
+    }
+
+fail:
+    return ret;
+}
+
 }  // namespace implementation
 }  // namespace V1_0/1
 }  // namespace power
diff --git a/Power.h b/Power.h
index 1c85a6d..868767f 100644
--- a/Power.h
+++ b/Power.h
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2017 The Android Open Source Project
- * Copyright (C) 2017 The LineageOS Project
+ * Copyright (C) 2017-2018 The LineageOS Project
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -23,6 +23,7 @@
 #else
 #include <android/hardware/power/1.1/IPower.h>
 #endif
+#include <vendor/lineage/power/1.0/ILineagePower.h>
 #include <hidl/MQDescriptor.h>
 #include <hidl/Status.h>
 #include <hardware/power.h>
@@ -44,13 +45,16 @@ using ::android::hardware::power::V1_0::IPower;
 #else
 using ::android::hardware::power::V1_1::IPower;
 #endif
+using ::vendor::lineage::power::V1_0::ILineagePower;
+using ::vendor::lineage::power::V1_0::LineageFeature;
 using ::android::hardware::Return;
 using ::android::hardware::Void;
 
-struct Power : public IPower {
+struct Power : public IPower, public ILineagePower {
     // Methods from ::android::hardware::power::V1_0::IPower follow.
 
     Power();
+    status_t registerAsSystemService();
 
     Return<void> setInteractive(bool interactive) override;
     Return<void> powerHint(PowerHint hint, int32_t data) override;
@@ -63,6 +67,9 @@ struct Power : public IPower {
     Return<void> powerHintAsync(PowerHint hint, int32_t data) override;
 #endif
 
+    // Methods from ::vendor::lineage::power::V1_0::ILineagePower follow.
+    Return<int32_t> getFeature(LineageFeature feature) override;
+
     // Methods from ::android::hidl::base::V1_0::IBase follow.
 
 };
diff --git a/hint-data.h b/hint-data.h
index 0cb54c2..61e064d 100644
--- a/hint-data.h
+++ b/hint-data.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012, 2013, 2015, 2017, The Linux Foundation. All rights reserved.
+ * Copyright (C) 2018 The LineageOS Project
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are
@@ -51,6 +52,7 @@
 
 #define VR_MODE_SUSTAINED_PERF_HINT    (0x1301)
 
+#define DEFAULT_PROFILE_HINT_ID         (0xFF00)
 
 struct hint_data {
     unsigned long hint_id; /* This is our key. */
diff --git a/power-common.h b/power-common.h
index 38191f6..601eb42 100644
--- a/power-common.h
+++ b/power-common.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2013, The Linux Foundation. All rights reserved.
+ * Copyright (C) 2018 The LineageOS Project
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are
@@ -41,6 +42,8 @@
 #define HINT_HANDLED (0)
 #define HINT_NONE (-1)
 
+#define ARRAY_SIZE(x) (sizeof((x))/sizeof((x)[0]))
+
 enum CPU_GOV_CHECK {
     CPU0 = 0,
     CPU1 = 1,
@@ -48,4 +51,12 @@ enum CPU_GOV_CHECK {
     CPU3 = 3
 };
 
+enum {
+    PROFILE_POWER_SAVE = 0,
+    PROFILE_BALANCED,
+    PROFILE_HIGH_PERFORMANCE,
+    PROFILE_BIAS_POWER,
+    PROFILE_BIAS_PERFORMANCE
+};
+
 #define UNUSED(x) UNUSED_ ## x __attribute__((__unused__))
diff --git a/power-helper.c b/power-helper.c
index c7aaf79..eeff2c8 100644
--- a/power-helper.c
+++ b/power-helper.c
@@ -1,7 +1,7 @@
 /*
  * Copyright (c) 2012-2017, The Linux Foundation. All rights reserved.
  * Copyright (C) 2017 The Android Open Source Project
- * Copyright (C) 2017 The LineageOS Project
+ * Copyright (C) 2017-2018 The LineageOS Project
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are
@@ -81,7 +81,6 @@
 #endif
 #endif
 
-#define ARRAY_SIZE(x) (sizeof((x))/sizeof((x)[0]))
 #define LINE_SIZE 128
 
 #ifdef LEGACY_STATS
@@ -319,11 +318,22 @@ void power_hint(power_hint_t hint, void *data)
         case POWER_HINT_VIDEO_DECODE:
             process_video_decode_hint(data);
         break;
+        case POWER_HINT_CPU_BOOST:
+            ALOGI("CPU boost power hint not handled in power_hint_override");
+        break;
+        case POWER_HINT_SET_PROFILE:
+            ALOGI("set profile power hint not handled in power_hint_override");
+        break;
         default:
         break;
     }
 }
 
+int get_number_of_profiles()
+{
+    return 0;
+}
+
 int __attribute__ ((weak)) set_interactive_override(int UNUSED(on))
 {
     return HINT_NONE;
diff --git a/power-helper.h b/power-helper.h
index 3d22353..e3be0f9 100644
--- a/power-helper.h
+++ b/power-helper.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2017, The Linux Foundation. All rights reserved.
+ * Copyright (C) 2017-2018 The LineageOS Project
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are
@@ -132,6 +133,7 @@ int extract_platform_stats(uint64_t *list);
 #ifndef V1_0_HAL
 int extract_wlan_stats(uint64_t *list);
 #endif
+int __attribute__ ((weak)) get_number_of_profiles();
 
 #ifdef __cplusplus
 }
diff --git a/service.cpp b/service.cpp
index ffb1f2a..d03c767 100644
--- a/service.cpp
+++ b/service.cpp
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2017 The Android Open Source Project
- * Copyright (C) 2017 The LineageOS Project
+ * Copyright (C) 2017-2018 The LineageOS Project
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -21,6 +21,8 @@
 #define LOG_TAG "android.hardware.power@1.1-service-qti"
 #endif
 
+// #define LOG_NDEBUG 0
+
 #include <android/log.h>
 #include <hidl/HidlTransportSupport.h>
 #include <hardware/power.h>
@@ -36,17 +38,16 @@ using android::hardware::joinRpcThreadpool;
 
 // Generated HIDL files
 #ifdef V1_0_HAL
-using android::hardware::power::V1_0::IPower;
 using android::hardware::power::V1_0::implementation::Power;
 #else
-using android::hardware::power::V1_1::IPower;
 using android::hardware::power::V1_1::implementation::Power;
 #endif
 
+
 int main() {
 
     status_t status;
-    android::sp<IPower> service = nullptr;
+    android::sp<Power> service = nullptr;
 
 #ifdef V1_0_HAL
     ALOGI("Power HAL Service 1.0 for QCOM is starting.");
@@ -63,7 +64,7 @@ int main() {
 
     configureRpcThreadpool(1, true /*callerWillJoin*/);
 
-    status = service->registerAsService();
+    status = service->registerAsSystemService();
     if (status != OK) {
         ALOGE("Could not register service for Power HAL Iface (%d).", status);
         goto shutdown;
-- 
2.7.4

